File: src/main/kotlin/dqualizer/dqexec/loadtest/mapper/k6/ParamsMapper.java
package dqualizer.dqexec.loadtest.mapper.k6;

import dqualizer.dqexec.config.PathConfig;
import dqualizer.dqexec.util.SafeFileReader;
import dqualizer.dqlang.archive.k6adapter.dqlang.k6.request.Request;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Map;
import java.util.Optional;

/**
 * Maps the request-parameter to Javascript-Code
 */
@Component
@RequiredArgsConstructor
public class ParamsMapper implements k6Mapper {

    private final PathConfig paths;
    private final SafeFileReader reader;

    @Override
    public String map(Request request) {
        Map<String, String> params = request.getParams();
        Optional<String> maybeReference = params.values().stream().findFirst();
        if(maybeReference.isEmpty()) return String.format("%sconst params = {}%s", newLine, newLine);

        String referencePath = paths.getResourcePath() + maybeReference.get();
        String paramsObject = reader.readFile(referencePath);

        return String.format("%sconst params = %s%s",
                newLine, paramsObject, newLine);
    }
}

File: src/main/kotlin/dqualizer/dqexec/loadtest/mapper/k6/ParamsMapper.kt
package dqualizer.dqexec.loadtest.mapper.k6

import dqualizer.dqexec.config.ResourcePaths
import io.github.dqualizer.dqlang.types.adapter.k6.request.Request
import org.springframework.stereotype.Component

/** Maps the request-parameter to Javascript-Code */
@Component
class ParamsMapper(private val resourcePaths: ResourcePaths) : K6Mapper {
  override fun map(request: Request): String {
    val params: Map<String, String> = request.params!!
    val path = params.values.first()

    if (path.isEmpty()) {
      return String.format("%sconst params = {}%s", K6Mapper.newLine, K6Mapper.newLine)
    }

    val paramsObject = resourcePaths.readResourceFile(path)
    return String.format(
      "%sconst params = %s%s",
      K6Mapper.newLine,
      paramsObject,
      K6Mapper.newLine,
    )
  }
}
